### ISSUE TITLE
Add support for Debian-based linux distros

### ISSUE THREAD

**Issue by CamoCatX** [rosenpass/rosenpass#45]
It would be nice if users of this distro would get the same benefits of other users like arch.

**Comment by wucke13** [comment]
No, not yet. @kpcyrd is there any chance that you might look into packaging this for debian? (@shibumi sugggested asking you when I asked him to do the arch packaging)

**Comment by kpcyrd** [comment]
In Debian every rust dependency needs to be packaged individually, you can use [cargo-debstatus](https://github.com/kpcyrd/cargo-debstatus) to check how many crates are still missing in Debian unstable:

![Untitled](https://user-images.githubusercontent.com/7763184/232209260-b469ad67-d103-47d4-a205-cd4b755b742a.png)

- `clap` 3 - the current version of clap in Debian is 4, but there's a `librust-clap-3-dev` package to satisfy the dependency for the time being
- `env_logger` 0.10 - Debian currently only has env_logger 0.9 because some dependencies are still missing in debian (is-terminal, io-lifetimes, rustix, linux-raw-sys). rosenpass would likely build with env_logger 0.9 too, so a maintainer might decide to relax this dependency
- `libsodium-sys-stable` - This crate is currently not in Debian yet, when uploading it ftp-masters might raise the question if both `libsodium-sys` and `libsodium-sys-stable` are needed inside of Debian (I don't know)
- `oqs-sys` - This crate is currently missing, but the C library `liboqs-dev` it depends on is already in Debian unstable, the `build-deps` crate needs to be packaged too
- `criterion` - since this is for benchmarking only and not needed for the build/tests it's usually simply patched away before uploading to Debian
- `test_bin` - this dependency is also missing, it's not strictly required since the binary could be built without it, but running tests is generally encouraged

The Debian Rust team allows you to package them without having a Debian Maintainer or Debian Developer status (somebody else is going to upload them for you), you can send merge requests here: https://salsa.debian.org/rust-team/debcargo-conf.

Most of the packaging is auto-generated from crates.io metadata, but Debian is also specifically about annotating copyright for every dependency and if the crate doesn't have a good description (missing, too long, too short) you need to write one yourself.

**Comment by wucke13** [comment]
Hi @kpcyrd, thank you for the elaboration! A few notes:

- `clap` v3 will be gone once #56 lands
- `libsodium-sys` is [kind-of deprecated](https://github.com/sodiumoxide/sodiumoxide), but some stuff of course still depends on it (which I think makes for an excellent argument why the debian policy of packaging everything themselves might not be the best idea and why the claim "then the distro maintainers can more easily track and update stuff might be worth questioning, but that's a topic for another issue :smile:)

Again thank you for providing the valuable information on how we would have to go forward with this. In the near-time future I will likely prefer to spend my time on improving the code, but at some point we will pursue this further. 

@CamoCatX just in case that this is not obvious: we provide statically linked rosenpass binaries in the release section on GitHub. These should work on _any_ x86_64 linux. So while not super convenient, there _is_ a way to use rosenpass on Debian based distros, just not to install it with `apt`.

**Comment by tranzystorek-io** [comment]
> * `oqs-sys` - This crate is currently missing, but the C library `liboqs-dev` it depends on is already in Debian unstable, the `build-deps` crate needs to be packaged too

https://github.com/open-quantum-safe/liboqs-rust/issues/190 is a slight roadbump here, as liboqs is always manually built and bundled into the binary

**Comment by koraa** [comment]
> > * `oqs-sys` - This crate is currently missing, but the C library `liboqs-dev` it depends on is already in Debian unstable, the `build-deps` crate needs to be packaged too
> 
> [open-quantum-safe/liboqs-rust#190](https://github.com/open-quantum-safe/liboqs-rust/issues/190) is a slight roadbump here, as liboqs is currently only manually built and bundled into the binary

That is a bit unfortunate; I don't think there is a timeline for when they are going to allow using the system's liboqs

**Comment by koraa** [comment]
@CamoCatX We didn't yet manage to provide a Debian package right? Why did you close the issue then?

**Comment by koraa** [comment]
Thank you for your contribution. I think Debian support is really important, its just hard to get done.

Where are you located. Want a t-shirt for your efforts?

**Comment by CamoCatX** [comment]
I appreciate the offer but I cannot reveal my location to the internet, I am trying to code Debian support because I believe in the cause of what this project is aimed to do.

**Comment by kpcyrd** [comment]
There's an effort to add pkg-config support to liboqs, so it can be easily picked up the rust bindings: https://github.com/open-quantum-safe/liboqs/issues/1434, https://github.com/open-quantum-safe/liboqs/pull/1435

In theory you could also attempt to introduce pkg-config support in the debian package for [liboqs](https://packages.debian.org/source/sid/liboqs) by opening an issue in the debian bug tracker (with a patch preferably), but I'd usually recommend to try to land these things upstream instead to avoid duplicated efforts. :)

**Comment by CamoCatX** [comment]
I would just like to say that this project will take awhile... There is much information for me to explore here. 

Best regards,
@CamoCatX
